crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -0,0 +1,395 @@
+
+/// Provider for timeline events prior to the current sync.
+pub trait PreviousEventsProvider: Send + Sync {
+    /// Returns the list of known timeline events, in sync order, for the given
+    /// room.
+    // TODO: return a reference or some kind of iterator
+    fn for_room(&self, room_id: &RoomId) -> Vec<SyncTimelineEvent>;
+}
+
+impl PreviousEventsProvider for () {
+    fn for_room(&self, _: &RoomId) -> Vec<SyncTimelineEvent> {
+        Vec::new()
+    }
+}
+
+#[instrument(skip_all, fields(room_id))]
+pub(crate) async fn compute_notifications(
+    client: &BaseClient,
+    changes: &StateChanges,
+    previous_events_provider: &dyn PreviousEventsProvider,
+    new_events: &[SyncTimelineEvent],
+    room_info: &mut RoomInfo,
+) -> Result<()> {
+    // Only apply the algorithm to encrypted rooms, since unencrypted rooms' unread
+    // notification counts ought to be properly computed by the server.
+    if !room_info.is_encrypted() {
+        return Ok(());
+    }
+
+    tracing::Span::current().record("room_id", display(&room_info.room_id));
+
+    let user_id = &client.session_meta().unwrap().user_id;
+    let prev_latest_receipt_event_id = room_info.latest_read_receipt_event_id.clone();
+
+    if let Some(receipt_event) = changes.receipts.get(room_info.room_id()) {
+        trace!("Got a new receipt event!");
+
+        // Find a private or public read receipt for the current user.
+        let mut receipt_event_id = None;
+        if let Some((event_id, receipt)) =
+            receipt_event.user_receipt(user_id, ReceiptType::ReadPrivate)
+        {
+            if receipt.thread == ReceiptThread::Unthreaded || receipt.thread == ReceiptThread::Main
+            {
+                receipt_event_id = Some(event_id.to_owned());
+            }
+        } else if let Some((event_id, receipt)) =
+            receipt_event.user_receipt(user_id, ReceiptType::Read)
+        {
+            if receipt.thread == ReceiptThread::Unthreaded || receipt.thread == ReceiptThread::Main
+            {
+                receipt_event_id = Some(event_id.to_owned());
+            }
+        }
+
+        if let Some(receipt_event_id) = receipt_event_id {
+            // We've found the id of an event to which the receipt attaches. The associated
+            // event may either come from the new batch of events associated to
+            // this sync, or it may live in the past timeline events we know
+            // about.
+
+            // First, save the event id as the latest one that has a read receipt.
+            room_info.latest_read_receipt_event_id = Some(receipt_event_id.clone());
+
+            // Try to find if the read receipts refers to an event from the current sync, to
+            // avoid searching the cached timeline events.
+            trace!("We got a new event with a read receipt: {receipt_event_id}. Search in new events...");
+            if find_and_count_events(&receipt_event_id, user_id, new_events.iter(), room_info) {
```

Comment:
See my comment below about `find_and_count_events` using `IntoIterator` to simplify this kind of call:

```rust
find_and_count_events(&receipt_event_id, user_id, new_events, room_info)
```